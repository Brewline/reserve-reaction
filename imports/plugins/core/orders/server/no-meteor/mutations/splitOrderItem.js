import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import Random from "@reactioncommerce/random";
import { Order as OrderSchema } from "/imports/collections/schemas";
import updateGroupStatusFromItemStatus from "../util/updateGroupStatusFromItemStatus";
import updateGroupTotals from "../util/updateGroupTotals";

/**
 * @name inputSchema
 * @private
 * @type {SimpleSchema}
 * @property {String} itemId The order item ID to split into two items
 * @property {Number} newItemQuantity The quantity that will be transferred to a new
 *   order item on the same fulfillment group.
 * @property {String} orderId The order ID
 */
const inputSchema = new SimpleSchema({
  itemId: String,
  newItemQuantity: {
    type: SimpleSchema.Integer,
    min: 1
  },
  orderId: String
});

/**
 * @method splitOrderItem
 * @summary Use this mutation to reduce the quantity of one item of an order and create
 *   a new item for the remaining quantity in the same fulfillment group, and with the
 *   same item status. You may want to do this if you are only able to partially fulfill
 *   the item order right now.
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Necessary input. See SimpleSchema
 * @return {Promise<Object>} Object with `order` property containing the created order
 *   and `newItemId` property set to the ID of the new item
 */
export default async function splitOrderItem(context, input) {
  inputSchema.validate(input);

  const {
    itemId,
    orderId,
    newItemQuantity
  } = input;

  const { appEvents, collections, isInternalCall, userHasPermission, userId } = context;
  const { Orders } = collections;

  // First verify that this order actually exists
  const order = await Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  // Allow split if the account has "orders" permission. When called internally by another
  // plugin, context.isInternalCall can be set to `true` to disable this check.
  if (!isInternalCall && !userHasPermission(["orders"], order.shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  const { billingAddress, cartId, currencyCode } = order;

  // Find and split the item
  let foundItem = false;
  const newItemId = Random.id();
  const orderSurcharges = [];
  const updatedGroups = await Promise.all(order.shipping.map(async (group) => {
    let itemToAdd;
    const updatedItems = group.items.map((item) => {
      if (item._id !== itemId) return item;

      if (item.quantity <= newItemQuantity) {
        throw new ReactionError("invalid-param", "quantity must be less than current item quantity");
      }

      foundItem = true;

      // The modified item to be created
      itemToAdd = {
        ...item,
        _id: newItemId,
        quantity: newItemQuantity,
        subtotal: item.price.amount * newItemQuantity
      };

      const updatedItemQuantity = item.quantity - newItemQuantity;
      const updatedItemSubtotal = item.price.amount * updatedItemQuantity;

      // The modified original item being split
      return {
        ...item,
        quantity: updatedItemQuantity,
        subtotal: updatedItemSubtotal
      };
    });

    if (!itemToAdd) return group;

    updatedItems.push(itemToAdd);

    // Create an updated group
    const updatedGroup = {
      ...group,
      // There is a convenience itemIds prop, so update that, too
      itemIds: updatedItems.map((item) => item._id),
      items: updatedItems,
      totalItemQuantity: updatedItems.reduce((sum, item) => sum + item.quantity, 0)
    };

    // Update group shipping, tax, totals, etc.
    const { groupSurcharges } = await updateGroupTotals(context, {
      billingAddress,
      cartId,
      currencyCode,
      discountTotal: updatedGroup.invoice.discounts,
      group: updatedGroup,
      orderId,
      selectedFulfillmentMethodId: updatedGroup.shipmentMethod._id
    });

    // Push all group surcharges to overall order surcharge array.
    // Currently, we do not save surcharges per group
    orderSurcharges.push(...groupSurcharges);

    // Ensure proper group status
    updateGroupStatusFromItemStatus(updatedGroup);

    return updatedGroup;
  }));

  // If we did not find any matching item ID while looping, something is wrong
  if (!foundItem) throw new ReactionError("not-found", "Order item not found");

  // We're now ready to actually update the database and emit events
  const modifier = {
    $set: {
      shipping: updatedGroups,
      surcharges: orderSurcharges,
      totalItemQuantity: updatedGroups.reduce((sum, group) => sum + group.totalItemQuantity, 0),
      updatedAt: new Date()
    }
  };

  OrderSchema.validate(modifier, { modifier: true });

  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
    { _id: orderId },
    modifier,
    { returnOriginal: false }
  );
  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");

  await appEvents.emit("afterOrderUpdate", {
    order: updatedOrder,
    updatedBy: userId
  });

  return { newItemId, order: updatedOrder };
}
