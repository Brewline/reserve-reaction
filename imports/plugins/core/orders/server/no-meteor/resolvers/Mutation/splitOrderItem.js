import { decodeOrderOpaqueId, decodeOrderItemOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/order";

/**
 * @name Mutation/splitOrderItem
 * @method
 * @memberof Payments/GraphQL
 * @summary resolver for the splitOrderItem GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.itemId - The ID of the item to split
 * @param {Number} args.input.newItemQuantity - The quantity that will be transferred to a new
 *   order item on the same fulfillment group.
 * @param {String} args.input.orderId - The order ID
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} SplitOrderItemPayload
 */
export default async function splitOrderItem(parentResult, { input }, context) {
  const {
    clientMutationId = null,
    newItemQuantity,
    itemId,
    orderId
  } = input;

  const { newItemId, order } = await context.mutations.splitOrderItem(context, {
    newItemQuantity,
    itemId: decodeOrderItemOpaqueId(itemId),
    orderId: decodeOrderOpaqueId(orderId)
  });

  return {
    clientMutationId,
    newItemId,
    order
  };
}
