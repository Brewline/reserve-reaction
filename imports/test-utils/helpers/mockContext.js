const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
