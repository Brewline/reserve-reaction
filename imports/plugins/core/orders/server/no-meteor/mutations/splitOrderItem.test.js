import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import {
  restore as restore$updateGroupTotals,
  rewire as rewire$updateGroupTotals
} from "../util/updateGroupTotals";
import splitOrderItem from "./splitOrderItem";

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(() => {
  restore$updateGroupTotals();
});

test("throws if orderId isn't supplied", async () => {
  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    newItemQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if itemId isn't supplied", async () => {
  await expect(splitOrderItem(mockContext, {
    orderId: "abc",
    newItemQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if newItemQuantity isn't supplied", async () => {
  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if newItemQuantity is 0", async () => {
  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 0
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if newItemQuantity is negative", async () => {
  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: -1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(null));

  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order item doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: [
          {
            _id: "xyz",
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if permission check fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: []
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders"], "SHOP_ID");
});

test("throws if newItemQuantity is equal to item quantity", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: [
          { _id: "abc", quantity: 2 }
        ]
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 2
  })).rejects.toThrowErrorMatchingSnapshot();
});


test("throws if newItemQuantity is greater than item quantity", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: [
          { _id: "abc", quantity: 1 }
        ]
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 2
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the database update fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        items: [
          Factory.OrderItem.makeOne({
            _id: "abc",
            quantity: 3,
            price: {
              amount: 1,
              currencyCode: "USD"
            },
            subtotal: 3,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  const mockUpdateGroupTotals = jest.fn().mockName("updateGroupTotals").mockReturnValue(Promise.resolve({ groupSurcharges: [] }));
  rewire$updateGroupTotals(mockUpdateGroupTotals);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 0
  }));

  await expect(splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("skips permission check if context.isInternalCall", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        items: [
          Factory.OrderItem.makeOne({
            _id: "abc",
            quantity: 2,
            price: {
              amount: 1,
              currencyCode: "USD"
            },
            subtotal: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  const mockUpdateGroupTotals = jest.fn().mockName("updateGroupTotals").mockReturnValue(Promise.resolve({ groupSurcharges: [] }));
  rewire$updateGroupTotals(mockUpdateGroupTotals);

  mockContext.isInternalCall = true;

  await splitOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    newItemQuantity: 1
  });

  delete mockContext.isInternalCall;

  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
});

test("splits an item", async () => {
  const item1 = Factory.OrderItem.makeOne({
    _id: "ITEM_1",
    price: {
      amount: 5,
      currencyCode: "USD"
    },
    quantity: 5,
    subtotal: 25,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const item2 = Factory.OrderItem.makeOne({
    _id: "ITEM_2",
    price: {
      amount: 5,
      currencyCode: "USD"
    },
    quantity: 1,
    subtotal: 5,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group = Factory.OrderFulfillmentGroup.makeOne({
    items: [item1, item2],
    itemIds: [item1._id, item2._id],
    totalItemQuantity: 6
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "ORDER_1",
    shipping: [group],
    shopId: "SHOP_ID",
    totalItemQuantity: 6,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  const mockUpdateGroupTotals = jest.fn().mockName("updateGroupTotals").mockReturnValue(Promise.resolve({ groupSurcharges: [] }));
  rewire$updateGroupTotals(mockUpdateGroupTotals);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await splitOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    newItemQuantity: 3
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "ORDER_1" },
    {
      $set: {
        shipping: [
          {
            ...group,
            items: [
              {
                ...item1,
                quantity: 2,
                subtotal: 10
              },
              item2,
              {
                ...item1,
                _id: jasmine.any(String),
                quantity: 3,
                subtotal: 15
              }
            ],
            itemIds: [item1._id, item2._id, jasmine.any(String)],
            totalItemQuantity: 6
          }
        ],
        surcharges: [],
        updatedAt: jasmine.any(Date),
        totalItemQuantity: 6
      }
    },
    { returnOriginal: false }
  );
});
