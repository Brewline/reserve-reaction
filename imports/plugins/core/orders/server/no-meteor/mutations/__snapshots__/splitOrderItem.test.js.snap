// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if itemId isn't supplied 1`] = `"Item ID is required"`;

exports[`throws if newItemQuantity is 0 1`] = `"New item quantity must be at least 1"`;

exports[`throws if newItemQuantity is equal to item quantity 1`] = `"quantity must be less than current item quantity"`;

exports[`throws if newItemQuantity is greater than item quantity 1`] = `"quantity must be less than current item quantity"`;

exports[`throws if newItemQuantity is negative 1`] = `"New item quantity must be at least 1"`;

exports[`throws if newItemQuantity isn't supplied 1`] = `"New item quantity is required"`;

exports[`throws if orderId isn't supplied 1`] = `"Order ID is required"`;

exports[`throws if permission check fails 1`] = `"Access Denied"`;

exports[`throws if the database update fails 1`] = `"Unable to update order"`;

exports[`throws if the order doesn't exist 1`] = `"Order not found"`;

exports[`throws if the order item doesn't exist 1`] = `"Order item not found"`;
